From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SugarcaneMC <SugarcaneMC@gmail.com>
Date: Fri, 23 Oct 2020 09:20:01 -0700
Subject: [PATCH] Utilities

From Yatopia
This patch includes all utilities required by the Sugarcane project and its patches.

Co-authored-by: Mykyta Komarnytskyy <nkomarn@hotmail.com>
Co-authored-by: Ivan Pekov <ivan@mrivanplays.com>

diff --git a/build.gradle.kts b/build.gradle.kts
index bfb9daee37f2ff9507729fc3fa1cae9570e41397..1e547cba9e8e1f7a1fa29f25859285db9d5f0c4d 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -47,6 +47,7 @@ dependencies {
     implementation("org.apache.logging.log4j:log4j-iostreams:2.14.1") // Paper
     implementation("org.apache.logging.log4j:log4j-api:2.14.1") // Paper
     implementation("org.ow2.asm:asm:9.1")
+    implementation("org.apache.commons:commons-rng-core:1.3")
     implementation("com.googlecode.json-simple:json-simple:1.1.1") {
         // This includes junit transitively for whatever reason
         isTransitive = false
diff --git a/src/main/java/org/sugarcanemc/sugarcane/server/util/Constants.java b/src/main/java/org/sugarcanemc/sugarcane/server/util/Constants.java
new file mode 100644
index 0000000000000000000000000000000000000000..b9a6b473a84f4310379f443c117a05108633d21c
--- /dev/null
+++ b/src/main/java/org/sugarcanemc/sugarcane/server/util/Constants.java
@@ -0,0 +1,7 @@
+package org.sugarcanemc.sugarcane.server.util;
+
+public class Constants {
+
+    public static final int[] EMPTY_ARRAY = new int[0];
+    public static final int[] ZERO_ARRAY = new int[]{0};
+}
diff --git a/src/main/java/org/sugarcanemc/sugarcane/server/util/FastRandom.java b/src/main/java/org/sugarcanemc/sugarcane/server/util/FastRandom.java
new file mode 100644
index 0000000000000000000000000000000000000000..2fa4f77632a37ebb1b1ab0b74f136de4447e27f5
--- /dev/null
+++ b/src/main/java/org/sugarcanemc/sugarcane/server/util/FastRandom.java
@@ -0,0 +1,64 @@
+package org.sugarcanemc.sugarcane.server.util;
+
+import org.apache.commons.rng.core.source64.XoRoShiRo128PlusPlus;
+
+import java.util.Random;
+import java.util.SplittableRandom;
+
+public class FastRandom extends Random {
+
+    private XoRoShiRo128PlusPlus random;
+
+    public FastRandom() {
+        super();
+        SplittableRandom randomseed = new SplittableRandom();
+        this.random = new XoRoShiRo128PlusPlus(randomseed.nextLong(), randomseed.nextLong());
+    }
+
+    public FastRandom(long seed) {
+        super(seed);
+        SplittableRandom randomseed = new SplittableRandom(seed);
+        this.random = new XoRoShiRo128PlusPlus(randomseed.nextLong(), randomseed.nextLong());
+    }
+
+    @Override
+    public boolean nextBoolean() {
+        return random.nextBoolean();
+    }
+
+    @Override
+    public int nextInt() {
+        return random.nextInt();
+    }
+
+    @Override
+    public float nextFloat() {
+        return (float) random.nextDouble();
+    }
+
+    @Override
+    public double nextDouble() {
+        return random.nextDouble();
+    }
+
+    @Override
+    public synchronized void setSeed(long seed) {
+        SplittableRandom randomseed = new SplittableRandom(seed);
+        this.random = new XoRoShiRo128PlusPlus(randomseed.nextLong(), randomseed.nextLong());
+    }
+
+    @Override
+    public void nextBytes(byte[] bytes) {
+        random.nextBytes(bytes);
+    }
+
+    @Override
+    public int nextInt(int bound) {
+        return random.nextInt(bound);
+    }
+
+    @Override
+    public long nextLong() {
+        return random.nextLong();
+    }
+}
diff --git a/src/main/java/org/sugarcanemc/sugarcane/server/util/TimeUtils.java b/src/main/java/org/sugarcanemc/sugarcane/server/util/TimeUtils.java
new file mode 100644
index 0000000000000000000000000000000000000000..96ff0318198b88fb34d06149bc338f1403aed57a
--- /dev/null
+++ b/src/main/java/org/sugarcanemc/sugarcane/server/util/TimeUtils.java
@@ -0,0 +1,27 @@
+package org.sugarcanemc.sugarcane.server.util;
+
+import java.util.concurrent.TimeUnit;
+
+public class TimeUtils {
+
+    public static String getFriendlyName(TimeUnit unit) {
+        switch (unit) {
+            case NANOSECONDS:
+                return "ns";
+            case MILLISECONDS:
+                return "ms";
+            case MICROSECONDS:
+                return "micros";
+            case SECONDS:
+                return "s";
+            case MINUTES:
+                return "m";
+            case DAYS:
+                return "d";
+            case HOURS:
+                return "h";
+            default:
+                throw new AssertionError();
+        }
+    }
+}
